### ISSUE TITLE
[RFC] Introduce ATOM as model implementation backend of SGLang

### ISSUE THREAD

**Issue by zejunchen-zejun** [ROCm/ATOM#295]
### Motivation
[ATOM](https://github.com/ROCm/ATOM) is a foundational component of AMD’s AI inference strategy. It is built by integrating optimizations from ROCm’s high-performance operator library [aiter](https://github.com/ROCm/aiter/tree/main) and high-performance communication library [mori](https://github.com/ROCm/mori) into model execution path. ATOM is used for new hardware launch and performance benchmark for customer's POC scenario. It can be used to serve as the model implementation backend for high-performance inference on AMD GPUs.

The primary purposes of adopting ATOM as the backend for SGLang model implementation are clearly outlined below, with a focus on supporting the smooth rollout of new technologies and the iterative improvement of related components.

1. With the official launch of AMD's new hardware products, ATOM will serve as a dedicated software platform for conducting comprehensive PoC testing, verifying the compatibility and stability of the new hardware in real-world application scenarios.
2. Prior to the full maturity of the new kernels and their associated API, all related prototypes will be developed, tested, and validated within the ATOM backend. This approach helps mitigate risks during the formal development phase and ensures a more robust final product.

In the initial stage, ATOM will function as a specialized PoC platform to complete the verification and iteration of new technologies. Once the new hardware and kernels reach the expected readiness level, all optimization results accumulated in ATOM will be promptly contributed to the SGLang, effectively facilitating the subsequent product incubation process.

SGLang and ATOM have distinct positioning: SGLang is built for product readiness, while ATOM serves as an experimental PoC platform dedicated to new technologies and platforms. This complementary collaboration creates a win-win scenario: SGLang reinforces its standing as the universal serving standard for LLM inference, and AMD users gain access to a high-performance, community-validated pathway for both production deployment and early adoption of new technologies. With this design, ATOM is engineered to empower the open-source community rather than fragment it. Our intent is to continue prioritizing in-tree native operator integration within SGLang, alongside sustained upstream contributions to drive the ecosystem forward.

### Focus
ATOM will focus on model-specific and performance-critical optimizations, including the following key capabilities:
- Cross-layer and cross-module fusion opportunities that span traditional layer boundaries, enabling more efficient computation
- KV Cache layouts specifically tailored to the requirements of the underlying kernel
- New optimized operators and kernel implementations designed to leverage the unique characteristics of AMD GPUs
- Advanced communication and runtime optimizations to enhance overall system performance

These optimizations are deeply rooted in model-level and kernel-level implementation details and do not interfere with any of SGLang's core design principles. ATOM is designed to reuse SGLang's excellent core components, allowing SGLang to maintain its position as a "first-class" platform for LLM serving on any high-performance hardware—including the ROCm ecosystem.

### Design
Here is a diagram showing how ATOM works as the model impl backend with SGLang. It uses existing model_impl backend registry (like [Transformers](https://github.com/huggingface/transformers) and [Mindspore)](https://gitee.com/mindspore/mindspore) and touches no core logic. It enables AMD GPU users to get better performance without changing their workflow—just add `--model-impl atom`.

<img width="779" height="528" alt="Image" src="https://github.com/user-attachments/assets/777d514d-7211-4cb8-ac38-86d8b206ec63" />

When frontend users launch SGLang with `--model-impl atom`, it will instantiate a thin wrapper model that delegates the heavy lifting to ATOM:
- SGLang creates `ATOMForCausalLM` for required model.
- The wrapper calls into ATOM’s plugin entrypoint when users have installed the ATOM. The entrypoint is shown as follow:
  - `model = atom.prepare_model(config=config, engine="sglang")`
- ATOM constructed an optimized model object (a callable) that implements the construction / forward / weight loading / logits computation methods, aligned with SGLang calling convention:
   - `__init__(...)`
   - `forward(...)`
   - `load_weights(...)`
   - `compute_logits(...)`

### Attention
SGLang provides official attention registry decorator `register_attention_backend`, which is a particularly clean design for enabling the alternative custom attention backend. ATOM leverages this registry mechanism and provide the required attention backend.

### Execution Flow
<img width="902" height="1150" alt="Image" src="https://github.com/user-attachments/assets/33babf5f-574b-4a15-a130-361b38bd36fd" />

### Functionality
Here is the accuracy value of vLLM with ATOM model impl backend:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     3|exact_match|↑  |0.8992|±  |0.0083|
|     |       |strict-match    |     3|exact_match|↑  |0.8772|±  |0.0090|

### Limitations
When running unsupported model on ATOM model impl backend, fallback to SGLang may hit AITER compatibility issue. In practice, trying to recover by switching model implementations back to SGLang native in the same environment can hit AITER skew risk, because ATOM and SGLang are compatible with different AITER versions.

### PRs
SGLang: https://github.com/sgl-project/sglang/pull/16944
ATOM: https://github.com/ROCm/ATOM/pull/296
